File: LeetCodeAlgos/Python/longestPalindrome.py
from collections import Counter
class Solution:
    def longestPalindrome(self, s: str) -> int:
        x, y, res = {}, list(s), 0

        for i in range(0, len(y)):
            if y[i] not in x.keys():
                x[y[i]] = 1
            else:
                x[y[i]] += 1
        for pairs in x.items():
            if pairs[1] % 2 == 0:
                res += pairs[1]
            elif pairs[1] != 1:
                res += pairs[1] // 2 * 2
                x[pairs[0]] %= 2
        if 1 in x.values():
            res += 1
        return res
    
    def longestPalindrome2(self, s: str) -> int:
        res = 0
        for x in Counter(s).values():
            res += x // 2 * 2
            if res%2==0 and x%2==1:
                res += 1
        return res

s = Solution()
print(s.longestPalindrome2("abccccdd"))
print(s.longestPalindrome2("a"))
print(s.longestPalindrome2("ccc"))
    

File: LeetCodeAlgos/Java/LongestPalindrome.java
// Given a string s which consists of lowercase or uppercase letters, return the length of the longest  palindrome that can be built with those letters.

// Letters are case sensitive, for example, "Aa" is not considered a palindrome.

import java.util.HashMap;

public class LongestPalindrome {
  public static void main(String[] args) {
    Solution s = new Solution();

    testLongestPalindrome(1, "abccccdd", 7, s);
    testLongestPalindrome(2, "a", 1, s);
    testLongestPalindrome(3, "ccc", 3, s);
  }

  public static void testLongestPalindrome(int testNum, String s, int expected, Solution sol) {
    int result = sol.longestPalindrome(s);

    System.out.println(String.format(
      "Test %d: %d / %d (%s)",
      testNum,
      result,
      expected,
      result == expected ? "PASS" : "FAIL"
    ));
  }
}

class Solution {
  public int longestPalindrome(String s) {
    HashMap<Character, Integer> count = new HashMap<Character, Integer>();
    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);
      if (count.containsKey(ch)) {
        count.put(ch, count.get(ch) + 1);
      } else {
        count.put(ch, 1);
      }
    }
    int len = 0;
    for (Integer val : count.values()){
      len += val / 2 * 2;
      if (len % 2 == 0 && val % 2 == 1) {
        len++;
      }
    }
    return len;
  }
}